### PR TITLE
feat(build): adding build system for hypercert abis, types, and contr…

### PR DESCRIPTION
**Pull request by brossetti1**
…act files

## Description

https://github.com/Network-Goods/hypercerts-protocol/issues/9

this WIP pr contains the beginning of a build system to release the hypercert contracts, types, and abis as its own package to be included in external apps that want to interact with contracts.

## Further Development

The package can be funneled down to only the contracts and types that need to be in final dist by removing or adding exports to the `src/index.ts` file and following the docs listed their.

I added a sort of cheap publish ci task - if you find it acceptable for now then`{hypercerts-npm-org}`would need to be replaced with the appropriate org name and `NPM_TOKEN` would need to be configured

## Environment Variables

- [ ] `NPM_TOKEN` would need to be configured

## Screenshots:

The final build package at this moment looks like the below image, of course if some of these files are not needed, they could be removed from `tsconfig.build.json` and `src/index.ts`

Screen Shot 2022-11-18 at 6 59 04 PM

### CODE REVIEW

File: .github/workflows/publish.yml [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+        if: steps.get-versions.outputs.package_version > steps.get-versions.outputs.npm_version
+      - run: yarn build
+        if: steps.get-versions.outputs.package_version > steps.get-versions.outputs.npm_version
+      - run: yarn test
+        if: steps.get-versions.outputs.package_version > steps.get-versions.outputs.npm_version
+      - name: publishing
+        if: steps.get-versions.outputs.package_version > steps.get-versions.outputs.npm_version
+        run: |
+          echo '//registry.npmjs.org/:_authToken=${{secrets.NPM_TOKEN}}' > .npmrc
+          npm publish --access public --scope {hypercerts-npm-org} --otp ${{ secrets.NPM_TOKEN }}
```
Comment (brossetti1) [inline .github/workflows/publish.yml:38]: if acceptable, `{hypercerts-npm-org}` would need to be replaced with the appropriate org

File: src/index.ts [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+export { ERC3525Upgradeable } from "./types/contracts/ERC3525Upgradeable";
+
+// Factories
+export { HyperCertMinter__factory as HyperCertMinterFactory } from "./types/factories/contracts/HyperCertMinter__factory";
+export { HyperCertSVG__factory as HyperCertSVGFactory } from "./types/factories/contracts/HyperCertSVG__factory";
+export { HyperCertMetadata__factory as HyperCertMetadataFactory } from "./types/factories/contracts/HyperCertMetadata.sol/HyperCertMetadata__factory";
+export { IHyperCertMinter__factory as IHyperCertMinterFactory } from "./types/factories/contracts/HyperCertMetadata.sol/IHyperCertMinter__factory";
+export { IHyperCertSVG__factory as IHyperCertSVGFactory } from "./types/factories/contracts/HyperCertMetadata.sol/IHyperCertSVG__factory";
+export { ERC3525SlotEnumerableUpgradeable__factory as ERC3525SlotEnumerableUpgradeableFactory } from "./types/factories/contracts/ERC3525SlotEnumerableUpgradeable__factory";
+export { ERC3525Upgradeable__factory as ERC3525UpgradeableFactory } from "./types/factories/contracts/ERC3525Upgradeable__factory";
```
Comment (brossetti1) [inline src/index.ts:42]: `index.ts` can be filtered down to only what is needed to export and as long as the `tsconfig.build.json` file is updated along with it, the final distribtuion will only contain the necessary files.

**Review by brossetti1 — COMMENTED** [review]
leaving some small comments

File: .github/workflows/publish.yml (hunk 1 above)
Comment (brossetti1) [reply .github/workflows/publish.yml:38]: ill probably come back through at some point and update this task to be more succicnt

**Comment by bitbeckers** [comment]
Added to project board to make sure we incorporate the publishing of contract types in the refactoring

**Comment by brossetti1** [comment]
closing - 🔥

**Pull request closed without merging**